src/common/webrequest_curl.cpp

```diff
@@ -0,0 +1,559 @@
+}
+
+wxString wxWebResponseCURL::GetHeader(const wxString& name) const
+{
+    wxWebRequestHeaderMap::const_iterator it = m_headers.find(name.Upper());
+    if ( it != m_headers.end() )
+        return it->second;
+    else
+        return wxString();
+}
+
+int wxWebResponseCURL::GetStatus() const
+{
+    long status = 0;
+    curl_easy_getinfo(GetHandle(), CURLINFO_RESPONSE_CODE, &status);
+    return status;
+}
+
+//
+// wxWebRequestCURL
+//
+
+static size_t wxCURLRead(char *buffer, size_t size, size_t nitems, void *userdata)
+{
+    if ( userdata )
+        return static_cast<wxWebRequestCURL*>(userdata)->ReadData(buffer, size * nitems);
+    else
+        return 0;
+}
+
+wxWebRequestCURL::wxWebRequestCURL(wxWebSession & session, int id, const wxString & url):
+    wxWebRequest(session, id)
+{
+    m_headers = session.GetHeaders();
+    m_headerList = NULL;
+
+    m_handle = curl_easy_init();
+    // Set error buffer to get more detailed CURL status
+    m_errorBuffer[0] = '\0';
+    curl_easy_setopt(m_handle, CURLOPT_ERRORBUFFER, m_errorBuffer);
+    // Set this request in the private pointer
+    curl_easy_setopt(m_handle, CURLOPT_PRIVATE, static_cast<void*>(this));
+    // Set URL to handle
+    curl_easy_setopt(m_handle, CURLOPT_URL, static_cast<const char*>(url.mb_str()));
+    // Set callback functions
+    curl_easy_setopt(m_handle, CURLOPT_WRITEFUNCTION, wxCURLWriteData);
+    curl_easy_setopt(m_handle, CURLOPT_HEADERFUNCTION, wxCURLHeader);
+    curl_easy_setopt(m_handle, CURLOPT_READFUNCTION, wxCURLRead);
+    curl_easy_setopt(m_handle, CURLOPT_READDATA, static_cast<void*>(this));
+    // Enable gzip, etc decompression
+    curl_easy_setopt(m_handle, CURLOPT_ACCEPT_ENCODING, "");
+    // Enable redirection handling
+    curl_easy_setopt(m_handle, CURLOPT_FOLLOWLOCATION, 1L);
+    // Limit redirect to HTTP
+    curl_easy_setopt(m_handle, CURLOPT_REDIR_PROTOCOLS,
+        CURLPROTO_HTTP | CURLPROTO_HTTPS);
+    // Enable all supported authentication methods
+    curl_easy_setopt(m_handle, CURLOPT_HTTPAUTH, CURLAUTH_ANY);
+    curl_easy_setopt(m_handle, CURLOPT_PROXYAUTH, CURLAUTH_ANY);
+}
+
+wxWebRequestCURL::~wxWebRequestCURL()
+{
+    DestroyHeaderList();
+
+    curl_easy_cleanup(m_handle);
+}
+
+void wxWebRequestCURL::Start()
+{
+    if ( GetState() != State_Idle )
+        return;
+
+    m_response.reset(new wxWebResponseCURL(*this));
+    m_response->Init();
+
+    if ( m_dataSize )
+    {
+        if ( m_method.empty() || m_method.IsSameAs("POST", false) )
+        {
+            curl_easy_setopt(m_handle, CURLOPT_POSTFIELDSIZE_LARGE,
+                static_cast<curl_off_t>(m_dataSize));
+            curl_easy_setopt(m_handle, CURLOPT_POST, 1L);
+        }
+        else if ( m_method.IsSameAs("PUT", false) )
+        {
+            curl_easy_setopt(m_handle, CURLOPT_UPLOAD, 1L);
+            curl_easy_setopt(m_handle, CURLOPT_INFILESIZE_LARGE,
+                static_cast<curl_off_t>(m_dataSize));
+        }
+    }
```

Comment:
I think there should be something (probably `wxFAIL`) here, i.e. if the method is explicitly set to something other than POST or PUT and data is provided.